autorag/nodes/queryexpansion/run.py

```diff
@@ -0,0 +1,174 @@
+        bm25_path = os.path.join(resources_dir, 'bm25.pkl')
+        assert bm25_path is not None, "bm25_path must be specified for using bm25 retrieval."
+        assert os.path.exists(bm25_path), f"bm25_path {bm25_path} does not exist. Please ingest first."
+
+        # run retrieval function
+        bm25_corpus = load_bm25_corpus(bm25_path)
+        original_bm25 = bm25.__wrapped__
+        ids, scores = original_bm25(queries=decomposed_queries, bm25_corpus=bm25_corpus, top_k=top_k)
+
+    elif retrieval_module_type == "vectordb":
+        # check if chroma_path and file exists
+        chroma_path = os.path.join(resources_dir, 'chroma')
+        assert chroma_path is not None, "chroma_path must be specified for using vectordb retrieval."
+        assert os.path.exists(chroma_path), f"chroma_path {chroma_path} does not exist. Please ingest first."
+
+        # set embedding model
+        embedding_model_str = module.module_param["embedding_model"]
+        if embedding_model_str in embedding_models:
+            embedding_model = embedding_models[embedding_model_str]
+        else:
+            logger.error(f"embedding_model_str {embedding_model_str} does not exist.")
+            raise KeyError(f"embedding_model_str {embedding_model_str} does not exist.")
+
+        # run retrieval function
+        chroma_collection = load_chroma_collection(db_path=chroma_path, collection_name=embedding_model_str)
+        original_vectordb = vectordb.__wrapped__
+        ids, scores = original_vectordb(queries=decomposed_queries, collection=chroma_collection,
+                                        embedding_model=embedding_model, top_k=top_k)
+    else:
+        raise ValueError(f"invalid f{retrieval_module} for using query_expansion_io decorator.")
+
+    return ids, scores
+
+
+def process_retrieval_module(retrieval_module, resources_dir, expanded_queries, top_k, data_dir):
+    ids, scores = retrieval_by_retrieval_module(retrieval_module=retrieval_module, resources_dir=resources_dir,
+                                                decomposed_queries=expanded_queries, top_k=top_k)
+    # fetch data from corpus_data
+    corpus_data = pd.read_parquet(os.path.join(data_dir, "corpus.parquet"))
+    contents = fetch_contents(corpus_data, ids)
+
+    # create retrieval df
+    retrieval_df = pd.DataFrame({
+        'retrieved_contents': contents,
+        'retrieved_ids': ids,
+        'retrieve_scores': scores,
+    })
+    return retrieval_df
+
+
+def module_best_retrieval(result_df: pd.DataFrame, top_k: int, retrieval_modules: List[Dict], resources_dir,
+                          retrieval_gt, data_dir, strategies):
+    # get expanded_queries to list
+    expanded_queries = result_df["expanded_queries"].tolist()
+
+    # get all combinations of retrieval modules
+    final_retrieval_modules = [item for retrieval_module in retrieval_modules
+                               for item in make_combinations(retrieval_module)]
```

Comment:
Do you consider using `explode` method from util.py?
Plus, it will occur some potential error, because 'module_type' key still in retrieval_modules. 
You must first pop module_type, find module callables (functions) from module.py, make combinations and explode it. 
Then you finally get list of module callables and list of module parameters.